### ISSUE TITLE
Use direct RPC streams to exchange CDC generations between nodes.

### ISSUE THREAD

**Issue by kbr-** [scylladb/scylla#7985]
Issue #7961 brought to our attention a problem with the current method of handling CDC generations.

Currently, when a new CDC generation is created (during bootstrap), the node which creates it inserts it into a distributed table: `system_distributed.cdc_generation_descriptions`. Then announces its timestamp through gossip.
Nodes attempt to retrieve the generation on certain events: restart, or when they notice that there is a new generation timestamp being gossiped. They retrieve it by reading the table.

The current implementation uses the following schema:
```
CREATE TABLE system_distributed.cdc_generation_descriptions (
    time timestamp PRIMARY KEY,
    description frozen<list<frozen<tuple<bigint, frozen<list<blob>>, tinyint>>>>,
    expired timestamp
) ...
```

The problem reported in #7961 is a large allocation reported by a node on restart, when i tries to retrieve a row from this table. The reason is that the entire generation is stored as a frozen collection - in a single cell. We store cells in continuous buffers. The collection can get quite large: asymptotically, its size is O(N * V * S), where N is the number of nodes, V is the number of tokens per node, and S is the number of shards on a node (assuming a homogeneous cluster; otherwise, we can take S to be the maximum of the number of shards over all nodes). With N = 10, V = 256, S = 38, the user reported a 2 MB allocation. With large clusters (say N = 100, S = 64) we can get allocations such as 34 MB.
Even if we used fragmented buffers or non-frozen collections problems would remain. The generation is stored in a single row and we don't yield when processing rows; hence the large row may (will) cause reactor stalls. Generally, we should follow our own advice and not use large collections.

In practice, these allocations and stalls should be harmless due to their rarity (generation changes are rare, and a stall or large allocation during node restart does not affect normal operation). But there are other problems stemming from the fact that we're using a distributed table:
- To read a generation from this table, its replicas must be available. This may not be true e.g. if we shutdown the entire cluster and we restart nodes one-by-one: the initial nodes may not be able to read any generation because the replicas that store it are still shutdown. They will keep retrying and reporting a warning message in the logs. They will eventually retrieve it (assuming that the user restarts all the nodes), but the initial warnings are still inelegant and unnecessary.
- We need to do special handling of single-node clusters. We use RF = 3 for the table and quorum reads/writes to make generations durable and available. But if the cluster is a single-node cluster, we cannot perform cl=quorum operations on the table.
- While CDC was still experimental, we had to change the schema of this table 3 times. Now comes the fourth time. Doing that is not easy since the table is distributed and used by nodes to communicate with each other. During the experimental period, we worked around the problem by simply using a new table with a different name and requiring user manual intervention if they enabled CDC (i.e. to call `nodetool checkAndRepairCdcStreams` after upgrading). Now we cannot afford such a thing. We need to introduce a new cluster feature before nodes start using the new table. Correctly exchanging generations during the upgrade period may be complex, because the node would have to consider both tables (keep using the old one until all nodes upgrade, then insert a generation into the new one). Also, introducing new cluster feature on each such change seems to be an overkill.
- To deal with the large allocation/stall problem, we need to partition the generation data - for example, by keeping the stream IDs for each vnode under a separate clustering row, or even having a separate row for each stream ID. But then, to insert a generation, we either have to use a large batch insert, or perform a sequence of inserts. The large batch may be too large - hence we may need to use a sequence of inserts. But that requires multiple round-trips, potentially resulting in larger than necessary latency and increasing the chance of failure. It does not make sense to partially insert a generation, so retrying would be necessary if we don't want to give up on the first small hiccup.
- With tables there are many moving parts and if one of these parts have a problem or behave unpredictably, the whole feature of CDC has a problem. That is the case #7961. When implementing CDC we assumed that Scylla tables... just work. But it turns out that collections - which appeared to be the most natural way to solve our problem - have this caveat that makes them improper for this use case.

Proposed solution
===
I propose a different, more natural way to solve the problem. When a node needs to retrieve a piece of information from another node, it can simply call that node. Adding a table in the middle - requiring the other node to insert the information to a table, and having the retriever read from that table - only adds unnecessary accidental complexity.

To send generations between nodes we can use direct RPC streams. Node A sees in gossiper that node B announces a new generation timestamp. Then node A calls node B, requesting it to stream the generation data. Implementation details below.

Storing generations
---

Currently we store generations in a distributed table. If we move away from this, they still need to be stored somewhere. I propose that each node has its own _local table_ to store generations.

This allows the node to send a generation to whoever calls and restore CDC data structures on restart without contacting any particular set of replicas.

The local table could have a schema e.g. as follows:
```
CREATE TABLE system.cdc_generations (
    timestamp timestamp,
    range_end bigint,
    ignore_msb tinyint,
    ok boolean static,
    expired boolean static,
    streams list<blob>,
    PRIMARY KEY (timestamp, vnode_end)
) ...
```
The sets of streams would be only local to a vnode, and their sizes would be bounded by the maximum number of shards on a node, so their sizes would be reasonable.

Generations would be inserted as follows:
1. iterate over all token ranges in the generation and insert their streams, each as a single row
2. update the `ok` column to `true`
3. flush

The static `ok` column is there to handle failures in the middle of inserting generations.
Still, such failures should be extremely rare, as we're dealing with a local table here and the entire procedure will be fast.

If necessary, changing the storage format is much easier with local tables than with distributed tables. The node can, for example, rewrite the table after restarting with an upgraded version.

By the way - we only ever need memtables for the insertion procedure. We also don't need to store anything in the cache. We can always read the table with cache bypass. The reads are rare and are required only for pulls and creating internal CDC data structures at restart.

Versioning generation formats
---

It may be sometimes necessary to change the serialization format or the contents of generations themselves. This introduces a slight challenge, since during rolling upgrade nodes may be using different formats, not being able to exchange generations.

To deal with this, we can introduce a version number which is always the first thing sent as the response to a generation pull request. Suppose node A wants to pull a generation from node B. Each node knows the version of the format that it currently understands. Node A first pulls B's version. If versions match there's no problem. Otherwise, A can simply abandon the pull. Reasoning:
- if A's version is greater, node B has yet to upgrade. Thus, node B shall eventually be restarted. Then A will attempt the pull again since it will notice that node B's application state changes. Or it will manage to pull that generation from a different node before node B upgrades; it doesn't matter from which node it eventually pulls.
- if B's version is greater, node A has yet to upgrade. Thus A will eventually be restarted and then it will retry the pull.

Introducing new generations
---

The procedure of introducing new generations would be similar to today's; we need only slight adjustments for the new method of storing generations.
1. A node decides that it needs to introduce a new generation. Today there are 2 ways: either through bootstrapping, or through `nodetool checkAndRepairCdcStreams`.
2. The node constructs the generation (finding stream IDs for each vnode/shard etc.) and inserts it to its local table as previously described.
3. The node updates its gossip application state with the generation's timestamp. We continue using the existing application state (`CDC_STREAMS_TIMESTAMP`).

Reacting to generation changes and retrieval
---

1. Each node observes its gossip state.
2. When a node A notices a new CDC generation timestamp, it randomly chooses a node from the set of nodes that gossip this timestamp. This is done to balance the load of pulling generations. Say the chosen node is B.
3. The node performs a RPC to B.
4. B reads the generation from its local table and streams it to A.
5. A performs the version check. If versions mismatch it abandons the pull. It remembers in memory not to attempt any more pulls from this node.
6. A inserts the generation to its local table and starts gossiping its timestamp. It also updates the in-memory CDC data structures used to perform CDC log writes with the new generation.

If the pull fails in the middle for whatever reason, A may retry (if it keeps seeing that B is alive).

Upgrade procedure from the current method
---

The new method is obviously incompatible with the current one. The new method introduces explicit versioning designed specifically for the problem at hand, making upgrades easy; but when switching between methods, our only tool are cluster features.

We introduce a cluster feature indicating that each node understands the new generation management method.

A node starts issuing pull requests using the new method only after it notices that the cluster feature is enabled: this means that every node understands the new method so every node is able to to respond to the pull request.
Note that a node may receive such a pull request _before_ it notices that the cluster feature is enabled. That is fine, however: because the pull request happened, it means that the node making the request saw the feature, which means that the receiver understands the new method as well.

Before the cluster feature is observed to be enabled on any node, we must ensure that every node can respond to generation pull requests. Thus, each node, before it joins the cluster feature, ensures that the generation it is currently gossiping is inserted to the new local table.

Until an upgraded node observes the feature to be enabled, it must keep using the old pulling method (i.e. using a distributed table).

An upgraded node may join the feature as soon as it can. _However!_ If the feature is enabled and the node gossips any generation timestamp, it must ensure that it stores that generation in the new local table (so it can answer pulls - they can start coming immediately after it joins the feature). Thus:
- if the node does not gossip any timestamp, it can simply join the feature.
- If it gossips a generation timestamp, it means it was able to retrieve the generation from distributed tables. Since its upgraded, before it started gossiping it, it could insert the generation to the new local table.

All in all, the best way to ensure that joining the feature is safe is to simply insert a generation to the new local table before a node starts gossiping that generation's timestamp.

As for deciding whether to gossip a timestamp, we do what we do today: when the node notices a new timestamp in gossip (in particular: after it restarts), it tries to retrieve the generation. Now:
- if the feature is already enabled, it pulls with the new method.
- Otherwise, it tries to read it from the old distributed table, if the table exists. It keeps retrying that until it gives up (this may happen if an "exotic" error occurs, or a new higher-timestamped generation appears), or it notices that the feature is enabled, in which case it switches to the new method.

After a successful pull (using either method), the node inserts the generation to the new local table and starts gossiping its timestamp.

One corner case is: there was no timestamp being gossiped. An upgraded node does not attempt to create new generations until it notices that the feature is enabled. After it does, and there are no timestamps being gossiped, the node creates a new CDC generation (this will happen, for example, if one jumps versions during upgrade).

Other notes?
---

It should be clear that serialization/deserialization of generations sent through the RPC streams should operate with fragmented buffers and yield when necessary. The implementation must be tested how it behaves with large generations - i.e. that it doesn't result in stalls or large allocations.

cc @haaawk @slivne @avikivity @elcallio @nyh @piodul

**Comment by haaawk** [comment]
The solution sounds reasonable to me in general. I have few questions:
1. What happens when a node that gossips generation timestamp is down before any other node manages to pull the new generation?
2. Why do we need ok static column when we gossip generation timestamp only after the local table was fully persisted? If the node proposing new generation crashes/restarts in the middle of persisting the table it's easy to check whether the last persisted generation is correct, no?
3. I don't think reasoning in `Versioning generation formats` is correct. Cluster has to keep working during rolling upgrade so nodes can't just ignore generations. The idea of versions is good we just need to make newer nodes work with older nodes when we introduce new version.

**Comment by haaawk** [comment]
@avikivity @tgrabiec @gleb-cloudius Could you please have a look and let us know whether you see any issues with the proposed solution. It would be marvellous if we could spot any problems now before @kbr- spends a week implementing this. Thanks a lot.

**Comment by gleb-cloudius** [comment]
I am not sure what exact requirement the cdc has in term of when the new
record has to be seen by others, but one thing come to mind: with
proposed approach each node will store all teh information (unlike now
where it is distributed between multiple nodes). How large the whole
data can be if one record now can reach 32M?

--
			Gleb.

**Comment by kbr-** [comment]
@haaawk 

> What happens when a node that gossips generation timestamp is down before any other node manages to pull the new generation?

And the timestamp-gossiping node never gets back up?
Then the generation will be forgotten. No one will actually know about it - except the first node, but it doesn't matter, since it's down.

Well, actually, there might be a problem... if the timestamp-gossiping node manages to insert the generation to the *client-facing table*.  Then we'll have a problem, because no node will actually use that generation for making CDC log writes, but clients will think that they do (since they pull it from the client-facing table) - so clients will be reading from that new generation, when they should be reading from the previous one.

Unfortunately... the current solution also has this problem. A bootstrapping node can first insert a generation into the internal table, then start gossiping it, then notice that the generation is in gossip (it notices its own gossip state) and update the client-facing table (as any node would), then crash before the timestamp reaches any other node.
We need to make this part fault-tolerant.

A solution would be to only insert the generation into the client-facing table only after we notice that some minimal number of nodes have learned of it, so we are certain that generation will propagate fully.

> Why do we need ok static column when we gossip generation timestamp only after the local table was fully persisted? If the node proposing new generation crashes/restarts in the middle of persisting the table it's easy to check whether the last persisted generation is correct, no?

You may be right on the first part. On the second part - we don't even need to check anything; such a failed generation will simply be ignored. A generation is only committed after the node starts gossiping it (and the timestamp being gossiped is currently stored in `system.cdc_local`). So we don't need an additional `ok` column.

> I don't think reasoning in Versioning generation formats is correct. Cluster has to keep working during rolling upgrade so nodes can't just ignore generations. The idea of versions is good we just need to make newer nodes work with older nodes when we introduce new version.

They work together by ignoring each other. Nodes won't ignore generations altogether: they will pull them from other nodes which they understand.

Now, if a node sees another node which has a newer generation, and it can't pull this generation from that node, or from any other node, then one of the following happened:
- that generation must have been created _in the middle of an upgrade_, i.e. the user bootstrapped a node before the entire cluster was upgraded; if the node uses a new version then old-version nodes won't be able to pull this new generation, and if the node uses an old version then new-version nodes won't be able to pull.
- Or a node was bootstrapped with the wrong version (into a single-version cluster). Actually, this can be thought of as a special case of the previous case.

Not being able to pull generations will cause some nodes to refuse coordinating writes to CDC-enabled tables (because they don't know the streams; they only know the timestamp from gossip). This will enforce the user to finish the upgrade and be careful next time (not bootstrapping a node with a wrong version / in the middle of upgrade).

Well, we can actually prevent this with a simple check: before bootstrapping, a node would first check version-compatibility with other nodes. If it detects incompatibility, it won't bootstrap, enforcing the user to finish the upgrade first / choose the right version for the bootstrapping node.

Another method would be to handle the incompatibility somehow. For example, with each new version, there would need to come a method of upgrading and downgrading generations from/to the previous version. However, I'm not sure if it's a good idea. I'm imagining that sometimes the two formats may be completely incompatible (e.g. if the semantics/meaning of stream IDs changes completely, not just the serialization format or whatever). I'd rather go for a solution which prevents the user from shooting themselves in their feet.

---

@gleb-cloudius 

> I am not sure what exact requirement the cdc has in term of when the new
record has to be seen by others, but one thing come to mind: with
proposed approach each node will store all teh information (unlike now
where it is distributed between multiple nodes). How large the whole
data can be if one record now can reach 32M?

We assume that the new record will be observed within `ring_delay`. Same as for tokens.

The amount of data is theoretically unbounded. A new generation is created for each bootstrapped node.
However, generations eventually expire - assuming that user puts TTLs on their CDC log tables and new generations are created - and we can garbage-collect such expired generations, as they are not needed for anything. So practically it shouldn't grow so much... for example, assuming that:
1. the user bootstraps at most 10 nodes a day
2. the user uses the default TTL of 24h on their CDC log tables

then there would be around 10 generations at any given moment, i.e. ~320M (assuming a ~100 node cluster, ~64 shards each).

That said, we still need to implement the garbage collection mechanism... see https://github.com/scylladb/scylla/issues/7300.

**Comment by kbr-** [comment]
@avikivity @gleb-cloudius I'm starting to code this - can I have your blessing please?

**Comment by kbr-** [comment]
@avikivity @gleb-cloudius
P I N G U S
M A X I M U S

blessing or objection please

**Comment by gleb-cloudius** [comment]
Sounds complicated, but it's not like I have better solution, so OK.

--
			Gleb.

**Comment by kbr-** [comment]
Thanks @gleb-cloudius !
I know there will be objections when the code is ready, but that's expected

**Comment by kbr-** [comment]
One fact occurred to my mind: in the internal table we don't really need to keep any generations other than the last one (the one whose timestamp we are gossiping).

**Comment by avikivity** [comment]
It looks like you're reinventing distributed tables. Local tables are equivalent to distributed tables (with RF=ALL). RPC streaming is an equivalent of paging. 

It's true that distributed tables are difficult, but inventing a new solution will bring with it new difficulties. Staying with tables allows any improvements we bring to tables (raft) to benefit cdc generations.

**Comment by kbr-** [comment]
> It looks like you're reinventing distributed tables. Local tables are equivalent to distributed tables (with RF=ALL). RPC streaming is an equivalent of paging.

Conceptually, I'm solving two very simple problems:
1. Send a message from node A to node B. The message contains a CDC generation.
2. Have a node persist the last known CDC generation so it can reload it on restart.

Currently, to solve both of these problems, we're using a distributed table.
1. To send the message, node A first stores it in a distributed table. Then node B retrieves it from the table.
2. For persistence of CDC generations, we're piggybacking on the above. I.e. since the generation was sent by storing it in a table, it is persisted. On restart, the node can read it from the table.

But using a distributed table to solve these problems requires a lot of work to be done that must be done ONLY because we're using a distributed table. I.e. it creates a lot of accidental complexity. The problem itself is however very simple (its essential complexity is low). It's just that we're making it very complex by not using the right tool for the job. Because we have the hammer called distributed tables in our hands, our problem looks like a nail. Using a distributed table does not remove any difficulties. It only adds them.

So what's the accidental complexity that we can get rid of?
1. The table must have a schema. Schemas are hard. I don't need or want a schema. I just want to send a message between two nodes. But because we must have a schema, I need to come up with the correct schema to store the message, and then making changes to that is *hard*. In this case, the change is modifying the serialization format of the message. If we keep using tables, then to change the format of the message, we need to create an additional table with a different schema and introduce a cluster feature so nodes don't accidentally use the wrong table / schema when sending or retrieving the message. See the "Split CDC streams..." PR.
Now, to be fair, there is a way to use a distributed table such that changing the message format does not require a schema change/new table: we can store bytes in the table. Of course, because we're using a table, we can't store these bytes in a single cell - as we were painfully reminded of recently. So the table would have a bunch of clustered rows storing a part of our message as bytes. I'm sure no one wants that.
2. The table must have a replication strategy. To insert or read from it, we must contact some set of replicas. What should be the RF? What should be the CL? Answers to these questions are not obvious, but I still must answer them to get the solution right. Recall this fragment of db/system_distributed_keyspace.cc:
```
/* We want to make sure that writes/reads to/from cdc_generations and cdc_streams
 * are consistent: a read following an acknowledged write to the same partition should contact
 * at least one of the replicas that the write contacted.
 * Normally we would achieve that by always using CL = QUORUM,
 * but there's one special case when that's impossible: a single-node cluster. In that case we'll
 * use CL = ONE for writing the data, which will do the right thing -- saving the data in the only
 * possible replica. Until another node joins, reads will also use CL = ONE, retrieving the data
 * from the only existing replica.
 *
 * There is one case where queries wouldn't see the read: if we extend the single-node cluster
 * with two nodes without bootstrapping (so the data won't be streamed to new replicas),
 * and the admin forgets to run repair. Then QUORUM reads might contact only the two new nodes
 * and miss the written entry.
 *
 * Fortunately (aside from the fact that nodes shouldn't be joined without bootstrapping),
 * after the second node joins, it will propose a new CDC generation, so the old entry
 * that was written with CL=ONE won't be used by the cluster anymore. All nodes other than
 * the first one use QUORUM to make the write.
 *
 * And even if the old entry was still needed for some reason, by the time the third node joins,
 * the second node would have already fixed our issue by running read repair on the old entry.
 */
static db::consistency_level quorum_if_many(size_t num_token_owners) {
    return num_token_owners > 1 ? db::consistency_level::QUORUM : db::consistency_level::ONE;
}
```
Now, did I get this right? I hope so, because I spent a few days thinking hard about this.
But why do I have to deal with this BS if all I want to do is send a message between two nodes?

3. You mentioned that "RPC streaming is an equivalent of paging". Ok, assuming that I got the schema right and managed to correctly implement the "send message" part (i.e. insert the message to the table), then the "read message" part (retrieve it from the table) is "easy" due to paging. But how about the send part? Inserting the message to the table correctly and performantly is not easy at all. Recall what I had to do in the PR that modifies the schema of the user-facing table:
```
static future<std::vector<mutation>> get_cdc_streams_descriptions_v2_mutation(
        const database& db,
        db_clock::time_point time,
        const cdc::topology_description& desc) {
    auto s = db.find_schema(system_distributed_keyspace::NAME, system_distributed_keyspace::CDC_DESC_V2);

    auto ts = api::new_timestamp();
    std::vector<mutation> res;
    res.emplace_back(s, partition_key::from_singular(*s, time));
    size_t size_estimate = 0;
    for (auto& e : desc.entries()) {
        // We want to keep each mutation below ~1 MB.
        if (size_estimate >= 1000 * 1000) {
            res.emplace_back(s, partition_key::from_singular(*s, time));
            size_estimate = 0;
        }

        set_type_impl::native_type streams;
        streams.reserve(e.streams.size());
        for (auto& stream : e.streams) {
            streams.push_back(data_value(stream.to_bytes()));
        }

        // We estimate 20 bytes per stream ID.
        // Stream IDs themselves weigh 16 bytes each (2 * sizeof(int64_t))
        // but there's metadata to be taken into account.
        // It has been verified experimentally that 20 bytes per stream ID is a good estimate.
        size_estimate += e.streams.size() * 20;
        res.back().set_cell(clustering_key::from_singular(*s, dht::token::to_int64(e.token_range_end)),
                to_bytes("streams"), make_set_value(cdc_streams_set_type, std::move(streams)), ts);

        co_await make_ready_future<>(); // maybe yield
    }

    co_return res;
}
```
and then:
```
    auto ms = co_await get_cdc_streams_descriptions_v2_mutation(_qp.db(), time, desc);
    co_await max_concurrent_for_each(ms, 20, [&] (mutation& m) -> future<> {
        // We use the storage_proxy::mutate API since CQL is not the best for handling large batches.
        co_await _sp.mutate(
            { std::move(m) },
            quorum_if_many(ctx.num_token_owners),
            db::timeout_clock::now() + 10s,
            nullptr, // trace_state
            empty_service_permit(),
            false // raw_counters
        );
    });
```
Why do I need to construct mutations? What should be their sizes? Why do I have to deal with all this BS if I just want to send a message?

---

Now, compare how easy it is to send a CDC generation using RPC streaming:
```
static future<> push_generation(
        cdc::topology_description generation,
        rpc::sink<cdc::token_range_description> generation_sink) {
    for (auto& e : std::move(*gen).entries()) {
        co_await generation_sink(std::move(e));
    }
}

static future<cdc::topology_description> pull_generation(
        rpc::source<cdc::token_range_description> generation_source) {
    std::vector<token_range_description> entries;
    while (auto frag_opt = co_await generation_source()) {
        auto [frag] = std::move(*frag_opt);
        entries.push_back(frag);
    }
    co_return topology_description{std::move(entries)};
}
```

Now the RPC framework takes care of any necessary batching behind the scenes. The solution is elegant and efficient. How nicer can it get?

---

As for the second problem - persisting the generation so a node can reload it on restart - by removing the distributed table, I need to solve this problem in a different way. So in that sense, using a distributed table does remove some difficulty. But this problem has an obvious solution that is perfectly suited for us - use a local table.
I don't need to think about replication. Yes, I still need to come up with a schema for this table - but fortunately, it's much easier because this schema is local to the node and doesn't need to be synchronized. Upgrades are easy because a node can simply do a local rewrite on restart. There is no need for inter-node synchronization.

**Comment by avikivity** [comment]
> > It looks like you're reinventing distributed tables. Local tables are equivalent to distributed tables (with RF=ALL). RPC streaming is an equivalent of paging.
> 
> Conceptually, I'm solving two very simple problems:
> 
> 1. Send a message from node A to node B. The message contains a CDC generation.
> 2. Have a node persist the last known CDC generation so it can reload it on restart.
> 
> Currently, to solve both of these problems, we're using a distributed table.
> 
> 1. To send the message, node A first stores it in a distributed table. Then node B retrieves it from the table.
> 2. For persistence of CDC generations, we're piggybacking on the above. I.e. since the generation was sent by storing it in a table, it is persisted. On restart, the node can read it from the table.
> 
> But using a distributed table to solve these problems requires a lot of work to be done that must be done ONLY because we're using a distributed table. I.e. it creates a lot of accidental complexity. The problem itself is however very simple (its essential complexity is low). It's just that we're making it very complex by not using the right tool for the job. Because we have the hammer called distributed tables in our hands, our problem looks like a nail. Using a distributed table does not remove any difficulties. It only adds them.
> 
> So what's the accidental complexity that we can get rid of?
> 
> 1. The table must have a schema. Schemas are hard. I don't need or want a schema. I just want to send a message between two nodes.


Messages also have schemas (IDL), and message schemas are just as hard has database schema. If you send the entire data set in one message, you will have a large allocation and latency problems. The IDL has to anticipate the addition of data due to new features. Finally, if you persist the message, you need some stable on-disk format.

> But because we must have a schema, I need to come up with the correct schema to store the message, and then making changes to that is _hard_.

The difficulty is inherent in making an always-on service with backwards compatibility. It's not due to the schema. Even if you stored everything in a json object (a la mongodb) you would still need to document how the object is laid out, and any changes would need to be backward compatible.

> In this case, the change is modifying the serialization format of the message. If we keep using tables, then to change the format of the message, we need to create an additional table with a different schema and introduce a cluster feature so nodes don't accidentally use the wrong table / schema when sending or retrieving the message. See the "Split CDC streams..." PR.

Let's say you had an RPC message to send the entire generation set. If you wanted to use RPC streams, you'd need a new verb that uses streams, and a cluster feature that tells you to use the new verb.

>    Now, to be fair, there is a way to use a distributed table such that changing the message format does not require a schema change/new table: we can store bytes in the table. Of course, because we're using a table, we can't store these bytes in a single cell - as we were painfully reminded of recently. So the table would have a bunch of clustered rows storing a part of our message as bytes. I'm sure no one wants that.

We don't! We also don't want to invisibly change the message format, because then we need a side channel to tell us what format to use, and whether all nodes understand the new format.

> 2. The table must have a replication strategy. To insert or read from it, we must contact some set of replicas. What should be the RF? What should be the CL? Answers to these questions are not obvious, but I still must answer them to get the solution right. 

Using RPC streams does not provide an answer. It just asks the question using different terms: do which node should I send the message? Can I tolerate failures? How do I read?

> Recall this fragment of db/system_distributed_keyspace.cc:
> 
> ```
> /* We want to make sure that writes/reads to/from cdc_generations and cdc_streams
>  * are consistent: a read following an acknowledged write to the same partition should contact
>  * at least one of the replicas that the write contacted.
>  * Normally we would achieve that by always using CL = QUORUM,
>  * but there's one special case when that's impossible: a single-node cluster. In that case we'll
>  * use CL = ONE for writing the data, which will do the right thing -- saving the data in the only
>  * possible replica. Until another node joins, reads will also use CL = ONE, retrieving the data
>  * from the only existing replica.
>  *
>  * There is one case where queries wouldn't see the read: if we extend the single-node cluster
>  * with two nodes without bootstrapping (so the data won't be streamed to new replicas),
>  * and the admin forgets to run repair. Then QUORUM reads might contact only the two new nodes
>  * and miss the written entry.
>  *
>  * Fortunately (aside from the fact that nodes shouldn't be joined without bootstrapping),
>  * after the second node joins, it will propose a new CDC generation, so the old entry
>  * that was written with CL=ONE won't be used by the cluster anymore. All nodes other than
>  * the first one use QUORUM to make the write.
>  *
>  * And even if the old entry was still needed for some reason, by the time the third node joins,
>  * the second node would have already fixed our issue by running read repair on the old entry.
>  */

That's a good question. Maybe the answer should have been to use the Everywhere replication strategy, with CL=ALL for writes and CL=LOCAL_ONE for reads. This is the equivalent to using local tables, plus broadcast for writes and local reads for reads.


> static db::consistency_level quorum_if_many(size_t num_token_owners) {
>     return num_token_owners > 1 ? db::consistency_level::QUORUM : db::consistency_level::ONE;
> }
> ```
> 
> Now, did I get this right? I hope so, because I spent a few days thinking hard about this.
> But why do I have to deal with this BS if all I want to do is send a message between two nodes?
> 

I don't think you want to send a message. You want all nodes to agree on a certain set of data. That's a harder problem, and the system should have a packaged solution for it, not force everyone to roll their own.



> 1. You mentioned that "RPC streaming is an equivalent of paging". Ok, assuming that I got the schema right and managed to correctly implement the "send message" part (i.e. insert the message to the table), then the "read message" part (retrieve it from the table) is "easy" due to paging. But how about the send part? Inserting the message to the table correctly and performantly is not easy at all. Recall what I had to do in the PR that modifies the schema of the user-facing table:
> 
> ```
> static future<std::vector<mutation>> get_cdc_streams_descriptions_v2_mutation(
>         const database& db,
>         db_clock::time_point time,
>         const cdc::topology_description& desc) {
>     auto s = db.find_schema(system_distributed_keyspace::NAME, system_distributed_keyspace::CDC_DESC_V2);
> 
>     auto ts = api::new_timestamp();
>     std::vector<mutation> res;
>     res.emplace_back(s, partition_key::from_singular(*s, time));
>     size_t size_estimate = 0;
>     for (auto& e : desc.entries()) {
>         // We want to keep each mutation below ~1 MB.
>         if (size_estimate >= 1000 * 1000) {
>             res.emplace_back(s, partition_key::from_singular(*s, time));
>             size_estimate = 0;
>         }
> 
>         set_type_impl::native_type streams;
>         streams.reserve(e.streams.size());
>         for (auto& stream : e.streams) {
>             streams.push_back(data_value(stream.to_bytes()));
>         }
> 
>         // We estimate 20 bytes per stream ID.
>         // Stream IDs themselves weigh 16 bytes each (2 * sizeof(int64_t))
>         // but there's metadata to be taken into account.
>         // It has been verified experimentally that 20 bytes per stream ID is a good estimate.
>         size_estimate += e.streams.size() * 20;
>         res.back().set_cell(clustering_key::from_singular(*s, dht::token::to_int64(e.token_range_end)),
>                 to_bytes("streams"), make_set_value(cdc_streams_set_type, std::move(streams)), ts);
> 
>         co_await make_ready_future<>(); // maybe yield
>     }
> 
>     co_return res;
> }
> ```
> 
> and then:
> 
> ```
>     auto ms = co_await get_cdc_streams_descriptions_v2_mutation(_qp.db(), time, desc);
>     co_await max_concurrent_for_each(ms, 20, [&] (mutation& m) -> future<> {
>         // We use the storage_proxy::mutate API since CQL is not the best for handling large batches.
>         co_await _sp.mutate(
>             { std::move(m) },
>             quorum_if_many(ctx.num_token_owners),
>             db::timeout_clock::now() + 10s,
>             nullptr, // trace_state
>             empty_service_permit(),
>             false // raw_counters
>         );
>     });
> ```
> 
> Why do I need to construct mutations? What should be their sizes? Why do I have to deal with all this BS if I just want to send a message?

I don't think you had to do it. I'm certain that for the rare occasion of rolling out a new CDC generation, using prepared statements would have been sufficient.

> 
> Now, compare how easy it is to send a CDC generation using RPC streaming:
> 
> ```
> static future<> push_generation(
>         cdc::topology_description generation,
>         rpc::sink<cdc::token_range_description> generation_sink) {
>     for (auto& e : std::move(*gen).entries()) {
>         co_await generation_sink(std::move(e));
>     }
> }
> 
> static future<cdc::topology_description> pull_generation(
>         rpc::source<cdc::token_range_description> generation_source) {
>     std::vector<token_range_description> entries;
>     while (auto frag_opt = co_await generation_source()) {
>         auto [frag] = std::move(*frag_opt);
>         entries.push_back(frag);
>     }
>     co_return topology_description{std::move(entries)};
> }
> ```
> 
> Now the RPC framework takes care of any necessary batching behind the scenes. The solution is elegant and efficient. How nicer can it get?

Maybe we should add a wrapper around prepared statements that make it similarly easy. Though note that RPC streaming also adds a sort of local transaction, which we don't have around a look of prepared statements.

> 
> As for the second problem - persisting the generation so a node can reload it on restart - by removing the distributed table, I need to solve this problem in a different way. So in that sense, using a distributed table does remove some difficulty. But this problem has an obvious solution that is perfectly suited for us - use a local table.
> I don't need to think about replication. Yes, I still need to come up with a schema for this table - but fortunately, it's much easier because this schema is local to the node and doesn't need to be synchronized. Upgrades are easy because a node can simply do a local rewrite on restart. There is no need for inter-node synchronization.

The database solves the same problem that you want to solve here, for its users. It's bad if it can't use the solutions we provide for users for our own use cases. In some cases it's impossible (bootstrapping the schema), but here I think it is merely inconvenient due to bad APIs.

**Comment by kbr-** [comment]
> Messages also have schemas (IDL), and message schemas are just as hard has database schema. If you send the entire data set in one message, you will have a large allocation and latency problems. The IDL has to anticipate the addition of data due to new features. Finally, if you persist the message, you need some stable on-disk format.

Messages have a schema - true, but with RPC I can easily agree on the schema on per-message basis before the message is sent. The entire solution is described in "Versioning generation formats" section.
1. Node A sends to node B the version of the "schema" that it wants to use for the RPC; the version number is always the first thing sent
2. If node B understands it, then OK, otherwise abandon the message sending (as described in the original post - that is OK)

With this the IDLs between versions don't need to be compatible at all.

And if I want, I can implement a translation step on node A after it learns that node B cannot understand the new version (or vice-versa) (but for this use case we don't even need any translation, as argued in the original post).

If I wanted to go with this idea with a table-based solution, I would need to have a separate mechanism - such as RPC - for the initial version exchange, and then use the correct table with the correct schema based on that exchange.

> The difficulty is inherent in making an always-on service with backwards compatibility. It's not due to the schema. Even if you stored everything in a json object (a la mongodb) you would still need to document how the object is laid out, and any changes would need to be backward compatible.

It depends on each use case. For this use case, we don't need backward compatibility of message "schemas" - as explained in the "Versioning generation formats" section.

> Let's say you had an RPC message to send the entire generation set. If you wanted to use RPC streams, you'd need a new verb that uses streams, and a cluster feature that tells you to use the new verb.

Not at all, I can keep using the same verb, I only need to ensure that the version number is the first thing that I send. Then - if both nodes agree on the version number - they know which "schema" will be used for the data coming through the RPC stream.

> We don't! We also don't want to invisibly change the message format, because then we need a side channel to tell us what format to use, and whether all nodes understand the new format.

What do you mean by "invisibly"?
With tables - whether I use a "proper schema" or store sequence of bytes in the table - I still need a side channel for nodes to agree on the message format (in case of storing bytes) or on which table to use (in case of using proper schemas).
With the solution I described, I can do it by using the same RPC verb for everything.

Actually, if I stored bytes in the table, I could also have a version number stored as the first thing, so I wouldn't need to use the side channel. The table would be used both for agreeing on the version of the message and sending the message.

> Using RPC streams does not provide an answer. It just asks the question using different terms: do which node should I send the message? Can I tolerate failures? How do I read?

I know answers to these questions by the problem definition.
1. Nodes are gossiping the generation timestamp
2. When node B learns that node A gossips a new generation timestamps, it tries to retrieve this generation. It knows that node A has the data.

Currently the retrieving of generations is done by reading from a table.
I want to change this solution to have node A send the data to node B using an RPC stream.

- to which node do I send the message: node A sends to node B, where A and B are defined above
- can I tolerate failures: I need node B to keep retrying the retrieval of the generation until it succeeds. Note that it can try contacting other nodes which are gossiping the same generation timestamp, because each of these nodes must have the data as well (otherwise  they wouldn't gossip the timestamp)
- how do I read: not sure what you mean; node B "reads" by pulling data from the RPC stream source

> That's a good question. Maybe the answer should have been to use the Everywhere replication strategy, with CL=ALL for writes and CL=LOCAL_ONE for reads. This is the equivalent to using local tables, plus broadcast for writes and local reads for reads.

Agreed. If we still want to go with tables, that's one aspect that I'd change compared to the previous solution.

> I don't think you want to send a message. You want all nodes to agree on a certain set of data. That's a harder problem, and the system should have a packaged solution for it, not force everyone to roll their own.

I want to broadcast a certain set of data through the cluster. One node creates the data, and other nodes simply retrieve it.
The high-level solution is already in-place:
1. The one node creates the data and gossips the identifier of this data (generation timestamp).
2. Nodes retrieve the data when they see it in gossip; after retrieving, they also gossip it
3. Thus the data spreads through the cluster like a virus

Now the question is how to do the "retrieve data" part. It can be done by sending a message from the node which gossips the identifier to the node which learns about the identifier.
Or it can be done by having the first node insert the data to a table, and then all other nodes retrieve it from this table after learning the identifier (the current solution). Here we have to answer questions about the table itself: about schema and replication.

Our experiences with using tables to solve this problem were mediocre so I'm suggesting an alternative.

I am not 100% convinced that using RPC streaming is superior. I really want you (or someone else) to convince me that tables are the better way. But considering all the complexity from using tables, I have strong reservations.

> I don't think you had to do it. I'm certain that for the rare occasion of rolling out a new CDC generation, using prepared statements would have been sufficient.

For the problem of creating a new CDC generation and broadcasting it through the cluster - I want (need) it to arrive on every node as quickly as possible, before anyone's clock moves past the generation's timestamp.

Assuming that the timestamp spreads through gossip within 30s (the "ring delay assumption" - which we already use for  tokens), if the nodes are not able to retrieve the data within 30s, they'll refuse to perform CDC log writes - because they know that there's a newer generation, but they don't know the stream IDs.

So efficiency of sending the message is important. It's also important to minimize the probability of failure in the middle of sending a message - so the faster it happens, the better.

> Maybe we should add a wrapper around prepared statements that make it similarly easy. Though note that RPC streaming also adds a sort of local transaction, which we don't have around a look of prepared statements.

You mean: I could provide a list of bindings, and these combined with a prepared statement would get translated to a sequence of mutations on the client side, which would be then sent to the server as a stream?

> The database solves the same problem that you want to solve here, for its users. It's bad if it can't use the solutions we provide for users for our own use cases. In some cases it's impossible (bootstrapping the schema), but here I think it is merely inconvenient due to bad APIs.

I just think (although I'm not 100% convinced - maybe about 70%) that tables are not the tool for this particular job. Similarly, consider the problem of repair: we don't use a table to send the mutation fragments by inserting them to the table and have the other side reading them from the table. We send them using an RPC stream. And I imagine that if we really wanted, we could send these mutation fragments using a table, however inconvenient that would be...

**Comment by avikivity** [comment]
> > Messages also have schemas (IDL), and message schemas are just as hard has database schema. If you send the entire data set in one message, you will have a large allocation and latency problems. The IDL has to anticipate the addition of data due to new features. Finally, if you persist the message, you need some stable on-disk format.
> 
> Messages have a schema - true, but with RPC I can easily agree on the schema on per-message basis before the message is sent. The entire solution is described in "Versioning generation formats" section.

Sure, but this increases the cost of the solution.

The question isn't whether a custom solution is better than a generic one. It usually is. The question is whether the problem merits the investment in a custom solution. For something as small and as rarely used as CDC generations, I think the answer is negative.

> 
> 1. Node A sends to node B the version of the "schema" that it wants to use for the RPC; the version number is always the first thing sent
> 2. If node B understands it, then OK, otherwise abandon the message sending (as described in the original post - that is OK)
> 
> With this the IDLs between versions don't need to be compatible at all.
> 
> And if I want, I can implement a translation step on node A after it learns that node B cannot understand the new version (or vice-versa) (but for this use case we don't even need any translation, as argued in the original post).
> 
> If I wanted to go with this idea with a table-based solution, I would need to have a separate mechanism - such as RPC - for the initial version exchange, and then use the correct table with the correct schema based on that exchange.

Well, with a table you usually ALTER the table if you want to modify the schema, you don't throw it away and get a new one.

> 
> > The difficulty is inherent in making an always-on service with backwards compatibility. It's not due to the schema. Even if you stored everything in a json object (a la mongodb) you would still need to document how the object is laid out, and any changes would need to be backward compatible.
> 
> It depends on each use case. For this use case, we don't need backward compatibility of message "schemas" - as explained in the "Versioning generation formats" section.
> 
> > Let's say you had an RPC message to send the entire generation set. If you wanted to use RPC streams, you'd need a new verb that uses streams, and a cluster feature that tells you to use the new verb.
> 
> Not at all, I can keep using the same verb, I only need to ensure that the version number is the first thing that I send. Then - if both nodes agree on the version number - they know which "schema" will be used for the data coming through the RPC stream.
> 
> > We don't! We also don't want to invisibly change the message format, because then we need a side channel to tell us what format to use, and whether all nodes understand the new format.
> 
> What do you mean by "invisibly"?

Without it being visible :)

> With tables - whether I use a "proper schema" or store sequence of bytes in the table - I still need a side channel for nodes to agree on the message format (in case of storing bytes) or on which table to use (in case of using proper schemas).

Usually, the "which table" question does not exist. You check which columns exist and use them.

> With the solution I described, I can do it by using the same RPC verb for everything.
> 
> Actually, if I stored bytes in the table, I could also have a version number stored as the first thing, so I wouldn't need to use the side channel. The table would be used both for agreeing on the version of the message and sending the message.
> 
> > Using RPC streams does not provide an answer. It just asks the question using different terms: do which node should I send the message? Can I tolerate failures? How do I read?
> 
> I know answers to these questions by the problem definition.
> 
> 1. Nodes are gossiping the generation timestamp
> 2. When node B learns that node A gossips a new generation timestamps, it tries to retrieve this generation. It knows that node A has the data.
> 
> Currently the retrieving of generations is done by reading from a table.
> I want to change this solution to have node A send the data to node B using an RPC stream.
> 
> * to which node do I send the message: node A sends to node B, where A and B are defined above
> * can I tolerate failures: I need node B to keep retrying the retrieval of the generation until it succeeds. Note that it can try contacting other nodes which are gossiping the same generation timestamp, because each of these nodes must have the data as well (otherwise  they wouldn't gossip the timestamp)
> * how do I read: not sure what you mean; node B "reads" by pulling data from the RPC stream source
> 
> > That's a good question. Maybe the answer should have been to use the Everywhere replication strategy, with CL=ALL for writes and CL=LOCAL_ONE for reads. This is the equivalent to using local tables, plus broadcast for writes and local reads for reads.
> 
> Agreed. If we still want to go with tables, that's one aspect that I'd change compared to the previous solution.
> 
> > I don't think you want to send a message. You want all nodes to agree on a certain set of data. That's a harder problem, and the system should have a packaged solution for it, not force everyone to roll their own.
> 
> I want to broadcast a certain set of data through the cluster. One node creates the data, and other nodes simply retrieve it.
> The high-level solution is already in-place:
> 
> 1. The one node creates the data and gossips the identifier of this data (generation timestamp).
> 2. Nodes retrieve the data when they see it in gossip; after retrieving, they also gossip it
> 3. Thus the data spreads through the cluster like a virus
> 
> Now the question is how to do the "retrieve data" part. It can be done by sending a message from the node which gossips the identifier to the node which learns about the identifier.
> Or it can be done by having the first node insert the data to a table, and then all other nodes retrieve it from this table after learning the identifier (the current solution). Here we have to answer questions about the table itself: about schema and replication.
> 
> Our experiences with using tables to solve this problem were mediocre so I'm suggesting an alternative.
> 

Instead, we should improve tables. The number of such problems is going to increase. Crafting a custom solution for each one will be expensive. Better to make the infrastructure more capable.

> I am not 100% convinced that using RPC streaming is superior. I really want you (or someone else) to convince me that tables are the better way. But considering all the complexity from using tables, I have strong reservations.
> 
> > I don't think you had to do it. I'm certain that for the rare occasion of rolling out a new CDC generation, using prepared statements would have been sufficient.
> 
> For the problem of creating a new CDC generation and broadcasting it through the cluster - I want (need) it to arrive on every node as quickly as possible, before anyone's clock moves past the generation's timestamp.

Our table don't have a way of notifying clients something changed, yet.

> 
> Assuming that the timestamp spreads through gossip within 30s (the "ring delay assumption" - which we already use for tokens), if the nodes are not able to retrieve the data within 30s, they'll refuse to perform CDC log writes - because they know that there's a newer generation, but they don't know the stream IDs.
> 
> So efficiency of sending the message is important. It's also important to minimize the probability of failure in the middle of sending a message - so the faster it happens, the better.

By efficiency I meant CPU and I/O cost. For generations, it doesn't matter if they take 10X the cpu and I/O - it's still small compared to the data. It's wasteful to improve it, because it doesn't change the amount of work the cluster has to do in a meaningful way.

> 
> > Maybe we should add a wrapper around prepared statements that make it similarly easy. Though note that RPC streaming also adds a sort of local transaction, which we don't have around a look of prepared statements.
> 
> You mean: I could provide a list of bindings, and these combined with a prepared statement would get translated to a sequence of mutations on the client side, which would be then sent to the server as a stream?

It doesn't have to be sent to the server as a stream, just make the concurrency control be automatic.

```c++
auto stmt = prepare(...);
auto sender = wrap_prepared_statement_for_concurrent_sending(stmt);
while (have more data) {
    co_await sender(pk, ck, col1, col2, col3);
}
co_await sender.close();
```

sender() will launch a write in parallel if we are under the concurrency limit, or wait for a write to complete if we are above it.

> 
> > The database solves the same problem that you want to solve here, for its users. It's bad if it can't use the solutions we provide for users for our own use cases. In some cases it's impossible (bootstrapping the schema), but here I think it is merely inconvenient due to bad APIs.
> 
> I just think (although I'm not 100% convinced - maybe about 70%) that tables are not the tool for this particular job. Similarly, consider the problem of repair: we don't use a table to send the mutation fragments by inserting them to the table and have the other side reading them from the table. We send them using an RPC stream. And I imagine that if we really wanted, we could send these mutation fragments using a table, however inconvenient that would be...

cdc generation is a control-plane operation. It doesn't have to be efficient (it can't be terrible, but it can be a little wasteful), it has to be maintainable, reliable, and not occupy too much of our limited attention.

Repair is a data-plane operation. It has to be well optimized, because its efficiency affects cluster performance directly.

**Comment by kbr-** [comment]
> Sure, but this increases the cost of the solution.

> The question isn't whether a custom solution is better than a generic one. It usually is. The question is whether the problem merits the investment in a custom solution. For something as small and as rarely used as CDC generations, I think the answer is negative.

It seems that you assume that the custom solution has  a "higher cost", is in some sense "harder", requires "more investment" than a table-based solution.

On the contrary: the solution proposed in original post is simple and a table-based solution is the harder one to craft properly.

The hard part of the problem is introducing a mechanism that would make the upgrade procedure easy when we later want to change the format of CDC generations. With the RPC-based proposed solution the mechanism is trivial. With a table-based solution - can't say at this point because I didn't invent and write it down yet, but with a 30s breadth-first search of the problem space in my brain I'd say that it's completely the opposite.

We could also use a table-based solution and ignore the upgrade problem. Then, whenever we need to modify the schema, we would need to come up with a custom upgrade / rewrite procedure, i.e. something akin to what's currently described in the "Upgrade procedure from the current method" section of the original post, or what we saw in the "Split CDC streams..." PR; but every time we need to do it again.

I'd say it's better to make a little investment and have a method which is "upgrade-aware", such as what is described in the original post (versioning formats), which makes later upgrades effortless. Of course I can't guarantee that every case would be covered, but many of them should be.

> Well, with a table you usually ALTER the table if you want to modify the schema, you don't throw it away and get a new one.

Getting a new table is easier than alter when we consider the fact that nodes keep using the table(s) while the change is undergoing.

Also note that currently the schemas of these internal distributed tables are simply hardcoded in system_distributed_keyspace.cc; we assume that they are constant and use `min_timestamp` for the schema mutations to make the synchronization easy. Nodes call `announce_migration` on startup; because the timestamps are equal, schema convergence is guaranteed. Crafting a proper solution which handles alters... I'm guessing it's going to take a lot of thinking and more investment than the proposed solution.

With a new table it's easier (but not easy): old nodes keep using the old table, new nodes keep using the old table until they see that everyone understands the new table (through a side channel), and that's when they start switching, which is safe at this point.

> Usually, the "which table" question does not exist. You check which columns exist and use them.

The existence of new columns (or a new table - using new columns vs using new table is equivalent; I'm assuming the new column names are disjoint from old column names) is not sufficient to say that it's safe to use them (the cluster may be in the middle of the upgrade, the data being still in the old columns). I need a side channel - such as a cluster feature - for nodes to agree if we can start using the new columns. And I need to copy the data from the old columns to the new ones.

> Crafting a custom solution for each one will be expensive.

For this problem, I'm pretty sure that using a table is the more complex and expensive thing to do.
Of course, the only way to be 100% sure would be to code both solutions and compare.

> By efficiency I meant CPU and I/O cost. For generations, it doesn't matter if they take 10X the cpu and I/O - it's still small compared to the data. It's wasteful to improve it, because it doesn't change the amount of work the cluster has to do in a meaningful way.

What's important here is that the operation completes quickly.

> sender() will launch a write in parallel if we are under the concurrency limit, or wait for a write to complete if we are above it.

What about batching? We need it in order to make the operation quick. We can't afford a separate round-trip for every row (again, it's not about resource usage, it's about the duration of the whole operation).

> cdc generation is a control-plane operation. It doesn't have to be efficient (it can't be terrible, but it can be a little wasteful), it has to be maintainable, reliable, and not occupy too much of our limited attention.

True.
Also, exchanging CDC generations must be fast.
And I argue that we can craft a custom solution which makes maintainability easy (especially upgrades), is reliable, and makes the operation fast. The solution is conceptually simple and with proper documentation everyone will be able to maintain it. The solution has a one-page high-level description; see the original post.
Using a hammer (tables) when we need a screwdriver is what makes the whole thing unnecessarily complex.

**Comment by haaawk** [comment]
@kbr- we should probably rename the issue to be less implementation centric since in the end we're gonna do internal table.

**Comment by kbr-** [comment]
I think we can just close it, the real "issue" is https://github.com/scylladb/scylla/issues/7961.
